File: src/walkway_navigation/scripts/ekf_monitor.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from nav_msgs.msg import Odometry
import math

class EKFMonitor(Node):
    def __init__(self):
        super().__init__('ekf_monitor')
        self.raw_sub = self.create_subscription(
            Odometry, '/odom', self.raw_odom_callback, 10)

        self.ekf_sub = self.create_subscription(
            Odometry, '/odometry/filtered', self.ekf_callback, 10)

        self.raw_x = 0.0
        self.raw_y = 0.0
        self.ekf_x = 0.0
        self.ekf_y = 0.0

        self.timer = self.create_timer(1.0, self.report_drift)

    def raw_odom_callback(self, msg):
        self.raw_x = msg.pose.pose.position.x 
        self.raw_y = msg.pose.pose.position.y 
    
    def ekf_callback(self, msg):
        self.ekf_x = msg.pose.pose.position.x
        self.ekf_y = msg.pose.pose.position.y

    def report_drift(self): 
        drift_x = abs(self.ekf_x - self.raw_x)
        drift_y = abs(self.ekf_y - self.raw_y)
        drift_2d = math.hypot(self.ekf_x - self.raw_x, self.ekf_y - self.raw_y)
        #self.get_logger().info(f'Current X drift between wheels and EKF: {drift_x:.4f} meters')
        #self.get_logger().info(f'Current Y drift between wheels and EKF: {drift_y:.4f} meters')
        self.get_logger().info(
            f'Drift -> X: {drift_x:.4f}m | Y: {drift_y:.4f}m'
        )

def main(args=None):
    rclpy.init(args=args)
    node = EKFMonitor()

    try:
        rclpy.spin(node)
    except (SystemExit, KeyboardInterrupt):
        node.get_logger().info('Spin broken. Cleaning up...')
    finally:
        node.destroy
        rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/walkway_navigation/src/ekf_monitor.cpp
#include "rclcpp/rclcpp.hpp"
#include "nav_msgs/msg/odometry.hpp"
#include <cmath>

class EKFMonitor : public rclcpp::Node {
    public:
    EKFMonitor() : Node("ekf_monitor"), raw_x_(0.0), raw_y_(0.0) {
        raw_sub_ = this->create_subscription<nav_msgs::msg::Odometry>(
            "/odom", 10, std::bind(&EKFMonitor::raw_odom_callback, this, std::placeholders::_1)
        );

        ekf_sub_ = this->create_subscription<nav_msgs::msg::Odometry>(
            "/odometry/filtered", 10, std::bind(&EKFMonitor::ekf_callback, this, std::placeholders::_1)
        );
    }

    private:
    void raw_odom_callback(const nav_msgs::msg::Odometry::SharedPtr msg) {
        raw_x_ = msg->pose.pose.position.x;
        raw_y_ = msg->pose.pose.position.y;
    }

    void ekf_callback(const nav_msgs::msg::Odometry::SharedPtr msg) const {
        double ekf_x = msg->pose.pose.position.x;
        double ekf_y = msg->pose.pose.position.y;
        double drift_x = std::abs(ekf_x - raw_x_);
        double drift_y = std::abs(ekf_y - raw_y_);
        RCLCPP_INFO(this->get_logger(), "Current X Drift between wheels and EKF: %.4f meters", drift_x);
        RCLCPP_INFO(this->get_logger(), "Current Y Drift between wheels and EKF: %.4f meters", drift_y);
    }

    rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr raw_sub_;
    rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr ekf_sub_;
    double raw_x_, raw_y_;
};

int main(int argc, char ** argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<EKFMonitor>();

    try {
        rclcpp::spin(node);
    }
    catch (const std::exception & e) {
        RCLCPP_ERROR(node->get_logger(), "Exception caught: %s", e.what());
    }

    RCLCPP_INFO(node->get_logger(), "Shutting down cleanly...");
    rclcpp::shutdown();
    return 0;
}